### PR TITLE
CSHARP-3117: Fix timeout connecting to server on Linux

### PR DESCRIPTION
**Pull request by vincentkam**
https://evergreen.mongodb.com/version/5ede94910305b9741f9e0041
I took Robert's commit: https://github.com/rstam/mongo-csharp-driver/commit/3c51ddc932e0019e7f677c88cbeecc263324e261
https://jira.mongodb.org/browse/CSHARP-3117

And tested against an Atlas shared tier cluster on WSL Ubuntu 18.04 like so:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net5.0</TargetFramework>
    <RootNamespace>driver_atlas_dns_test</RootNamespace>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="MongoDB.Driver" Version="2.11.0-beta1" />
  </ItemGroup>

</Project>
```

```c#
// Program.cs
        static void Main(string[] args)
        {
            var connectionString = "mongodb+srv://user:password@cluster0-8qo69.mongodb.net/admin?retryWrites=true&w=majority";
            var client = new MongoClient(connectionString);
            foreach (var db in client.ListDatabases().ToList()) Console.WriteLine(db);
        }
```
On Windows, the test app works as expected. On Linux, I took the same code in Program.cs and dumped it into our `TestConsoleApplication` project, removing the `net452` target (leaving only `netcoreapp2.1`.
Without the fix, we see the same exception as seen by the initial report:
```bash

vincent@Astorma:~/local-projects/mongo-csharp-driver/tests/MongoDB.Driver.TestConsoleApplication$ dotnet run --framework netcoreapp2.1

Unhandled Exception: System.TimeoutException: A timeout occured after 30000ms selecting a server using CompositeServerSelector{ Selectors = MongoDB.Driver.MongoClient+AreSessionsSupportedServerSelector, LatencyLimitingServerSelector{ AllowedLatencyRange = 00:00:00.0150000 } }. Client view of cluster state is { ClusterId : "1", ConnectionMode : "ReplicaSet", Type : "ReplicaSet", State : "Disconnected", Servers : [{ ServerId: "{ ClusterId : 1, EndPoint : "Unspecified/cluster0-shard-00-00-8qo69.mongodb.net:27017" }", EndPoint: "Unspecified/cluster0-shard-00-00-8qo69.mongodb.net:27017", ReasonChanged: "Heartbeat", State: "Disconnected", Type: "Unknown", HeartbeatException: "MongoDB.Driver.MongoConnectionException: An exception occurred while opening a connection to the server. ---> System.PlatformNotSupportedException: Sockets on this platform are invalid for use after a failed connection attempt.

```

After applying Robert's fix, I reran `TestConsoleApplication` and it successfully connected and dumped a list of the DB names.

### CODE REVIEW

**Comment by vincentkam** [comment]
Ready for review!

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

File: src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs [hunk 1]
```diff
@@ -181,7 +181,7 @@ private async Task ConnectAsync(Socket socket, EndPoint endPoint, CancellationTo
                 try
                 {
                     var dnsEndPoint = endPoint as DnsEndPoint;
-#if NETSTANDARD1_5
+#if NETSTANDARD1_5 || NETSTANDARD2_0
```
Comment (jyemin) [inline src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs:184]: Just curious why nothing needs to be done in the `Connect` method.

Comment (vincentkam) [reply src/MongoDB.Driver.Core/Core/Connections/TcpStreamFactory.cs:184]: Per https://github.com/dotnet/runtime/issues/24917
> This is only happening under specific circumstances:
[1] host is a hostname, such as "localhost", AND [2] KeepAlive is enabled using tcpClient.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true)

So we currently bypass this problem by resolving the host for the  .NET Standard version of our driver and then [calling connect](https://github.com/vincentkam/mongo-csharp-driver/pull/117/files#diff-c823181a4e53905ddcff62eccc5afca7R78) (ditto in the [async version](https://github.com/vincentkam/mongo-csharp-driver/pull/117/files#diff-c823181a4e53905ddcff62eccc5afca7L92)), thereby negating the first condition.

**Review by rstam — APPROVED** [review]
LGTM

**Comment by vincentkam** [comment]
Date amended and pushed

**Pull request merged**